lib/evaluate/check-call.cc

```diff
@@ -0,0 +1,199 @@
+              if (auto type{characteristics::GetTypeAndShape(*expr, context)}) {
+                object.type.IsCompatibleWith(context.messages(), *type);
+              } else {
+                // TODO
+              }
+            } else {
+              // TODO
+            }
+          },
+          [&](const characteristics::DummyProcedure &) {
+            // TODO check effective procedure compatibility
+          },
+          [&](const characteristics::AlternateReturn &) {
+            // TODO check alternate return
+          },
+      },
+      dummy.u);
+}
+
+static void RearrangeArguments(const characteristics::Procedure &proc,
+    ActualArguments &arg, parser::ContextualMessages &messages) {
+  CHECK(proc.HasExplicitInterface());
+  if (arg.size() < proc.dummyArguments.size()) {
+    arg.resize(proc.dummyArguments.size());
+  } else if (arg.size() > proc.dummyArguments.size()) {
+    messages.Say(
+        "%zd effective argument%s are passed in a reference to a procedure that expects only %zd"_err_en_US,
+        arg.size(), arg.size() == 1 ? "" : "s", proc.dummyArguments.size());
+  }
+  std::map<std::string, ActualArgument> kwArgs;
+  for (auto &x : arg) {
+    if (x.has_value()) {
+      if (x->keyword.has_value()) {
+        auto emplaced{
+            kwArgs.try_emplace(x->keyword->ToString(), std::move(*x))};
+        if (!emplaced.second) {
+          messages.Say(*x->keyword,
+              "Argument keyword '%s=' appears on more than one effective argument in this procedure reference"_err_en_US,
+              *x->keyword);
+        }
+        x.reset();
+      }
+    }
+  }
+  if (!kwArgs.empty()) {
+    int index{0};
+    for (const auto &dummy : proc.dummyArguments) {
+      if (!dummy.name.empty()) {
+        auto iter{kwArgs.find(dummy.name)};
+        if (iter != kwArgs.end()) {
+          ActualArgument &x{iter->second};
+          if (arg[index].has_value()) {
+            messages.Say(*x.keyword,
+                "Keyword argument '%s=' has already been specified positionally (#%d) in this procedure reference"_err_en_US,
+                *x.keyword, index + 1);
+          } else {
+            arg[index] = std::move(x);
+          }
+          kwArgs.erase(iter);
+        }
+      }
+      ++index;
+    }
+    for (auto &bad : kwArgs) {
+      ActualArgument &x{bad.second};
+      messages.Say(*x.keyword,
+          "Argument keyword '%s=' is not recognized for this procedure reference"_err_en_US,
+          *x.keyword);
+    }
+  }
+}
+
+void CheckArguments(const characteristics::Procedure &proc,
+    ActualArguments &arg, FoldingContext &context,
+    bool treatingExternalAsImplicit) {
+  parser::Messages buffer;
+  parser::ContextualMessages messages{context.messages().at(), &buffer};
+  FoldingContext localContext{context, messages};
+  if (proc.HasExplicitInterface()) {
+    RearrangeArguments(proc, arg, messages);
+    int index{0};
+    for (auto &x : arg) {
```

Comment:
As above, I don't think `arg` is a good name for a collection of arguments.
Given the presence of `dummy` in this function, renaming `x` to `actual` and `arg` to `actuals` or `actualArguments` would make it easier to understand.

Reply:
Renaming both to `actuals`.